<a href="https://colab.research.google.com/github/nurfnick/Numerical_Methods/blob/master/WeBWorK2HomeworkHelper.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# WeBWorK Homework Helper 2



## Number 1

The definition of the derivative, applied to $f(x) = 7^x$, results in the formula $f'(x) = k \cdot 7^x$ where
$$k = \lim_{h \rightarrow 0} \frac{7^h - 1}{h}$$ .

The value of the constant $k$ cannot be determined using the usual tricks for evaluating limits. Using your calculator, approximate the constant $k$, rounded to three significant figures.

In [ ]:
def quo(h):
  return (7**h-1)/h

testvalues =[1,.1,.01,.001,.0001]

for i in testvalues:
  print(quo(i))

6.0
2.148140440390669
1.9649663856459165
1.9478046608600863
1.9460994896514094


I want three sig figs, so I think the answer is 1.95 if I round up.  If I needed another sig fig, I would go ahead and try another lower power of ten.

## WeBWorK 7

Given the following table:

| x    | 0.0061     | 0.0062      | 0.0063     | 0.0064      | 0.0065      |
|------|------------|-------------|------------|-------------|-------------|
| f(x) | 0.54098533 | -0.87676671 | 0.99682327 | -0.73768343 | 0.091756934 |

Calculate the value of $f′(0.0063)$ to two places of accuracy.

To obtain more precise information about the value of f near 0.0063 enter a new increment value for x here 
0.0001
 and then press the Submit Answer button.

How will you tell when your increment is small enough to give you a good answer for the problem?

So I am going to use the forward difference and the backward difference.

In [ ]:
(-.73768343-0.99682327)/.0001

-17345.067

In [ ]:
(0.99682327+0.87676671)/.0001

18735.8998

So which value do you think is a good estimate for the derivative?  Neither is the correct answer!  I need to get a smaller increment, $h=0.0001$ is not cutting it!  I went back to WeBWorK and used $h=0.00001$.

| x    | 0.00628    | 0.00629    | 0.0063     | 0.00631   | 0.00632    |
|------|------------|------------|------------|-----------|------------|
| f(x) | 0.83357965 | 0.94536534 | 0.99682327 | 0.9852746 | 0.91203351 |

I'll again compute the forward and the backward difference.

In [ ]:
(0.9852746-0.99682327)/0.00001

-1154.8670000000038

In [ ]:
(0.99682327-0.94536534)/0.00001

5145.792999999998

Well dang it!  I'll change the step size to $h=0.000001$

| x    | 0.006298   | 0.006299 | 0.0063     | 0.006301   | 0.006302   |
|------|------------|----------|------------|------------|------------|
| f(x) | 0.99154422 | 0.9945   | 0.99682327 | 0.99851315 | 0.99956917 |

And again!

In [ ]:
(0.99851315-0.99682327)/.000001

1689.8799999999214

In [ ]:
(0.99682327-0.9945)/0.000001

2323.2699999999886

I am getting closer but still don't have two sig figs!  $h=0.0000001$

| x    | 0.0062998  | 0.0062999  | 0.0063     | 0.0063001  | 0.0063002  |
|------|------------|------------|------------|------------|------------|
| f(x) | 0.99640926 | 0.99661943 | 0.99682327 | 0.99702077 | 0.99721194 |

Last time hopefully!

In [ ]:
(0.99702077-0.99682327)/0.0000001

1974.999999999616

In [ ]:
(0.99682327-0.99661943)/.0000001

2038.400000000662

I promise, last time.  $h=0.00000001$

| x    | 0.00629998 | 0.00629999 | 0.0063     | 0.00630001 | 0.00630002 |
|------|------------|------------|------------|------------|------------|
| f(x) | 0.99678301 | 0.99680317 | 0.99682327 | 0.9968433  | 0.99686328 |

so


In [ ]:
(0.9968433-0.99682327)/0.00000001

2002.99999999487

In [ ]:
(0.99682327- 	0.99680317)/.00000001

2010.0000000078387

So I know the derivative is between 2010 and 2003.  I am only required two sig figs so 2000 will work as the answer!  

tl;dr Use a really small h to start on this problem!

## WeBWorK 11

The temperature adjusted for wind chill is a temperature that tells you how cold it feels, as a result of the combination of wind speed, w, in miles per hour, and actual temperature, T, as measured by a thermometer in degrees Fahrenheit. The table below shows the temperature adjusted for wind chill, f(w,T), as a function of w and T. The left column of the table gives the wind speed in miles per hour, the top row gives the actual temperature in degrees Fahrenheit, and the remaining entries give the temperature adjusted for wind chill in degrees Fahrenheit as a function of wind speed and actual temperature. 

| w\T | 25 | 20 | 15 | 10  | 5   | 0   | -5  | -10 |
|-----|----|----|----|-----|-----|-----|-----|-----|
| 5   | 19 | 13 | 7  | 1   | -5  | -11 | -16 | -22 |
| 10  | 15 | 9  | 3  | -4  | -10 | -16 | -22 | -28 |
| 15  | 13 | 6  | 0  | -7  | -13 | -19 | -26 | -32 |
| 20  | 11 | 4  | -2 | -9  | -15 | -22 | -29 | -35 |
| 25  | 9  | 3  | -4 | -11 | -17 | -24 | -31 | -37 |

 Estimate $f_w(10,10)$ and give a practical interpretation. $f_w(10,10)\approx$

If the wind is blowing mph and the actual temperature is degrees Fahrenheit, it will feel degrees Fahrenheit colder for each 1 mph increase in wind speed.


So I know that many of you have not taken multivariate calculus but the real world does not just have one variable!  To estimate partial derivatives, we will do the same as with one variable and holding the other values constant.  There are actually two estiates we can make here and I'll make them both.

$$f_w(10,10)\approx \frac{f(10+h,10)-f(10,10)}h
  $$

I can use $h=\pm5$

In [ ]:
# h is +5

print((-7+4)/5)

# h is -5

print( (1+4)/(-5))

-0.6
-1.0


## WeBWorK 23

Find a polynomial of the form
$$f(x)=ax^3+bx^2+cx+d$$
such that $f(0)=5$, $f(2)=6$, $f(3)=-6$, and $f(-4)=7$. 

I'll apply Lagrange Interpolation.

In [ ]:
def lagrangeCoefficient(inputArray,outputArray,entry,variable):
  lc = 1
  for i in range(len(inputArray)):
    if i == entry:
      continue
    else:
      lc = lc*(variable - inputArray[i])/(inputArray[entry]-inputArray[i])
  return lc

def lagrangeInterpolator(inputArray, outputArray, variable):
  li = 0
  for i in range(len(inputArray)):
    li = li + lagrangeCoefficient(inputArray,outputArray,i,variable)*outputArray[i]
  return li

In [ ]:
input = [0,2,3,-4]
output = [5,6,-6,7]

lagrangeInterpolator(input,output,0) #this is d


5.0

Again though the issue I am going to have is that I cannot do this symbolically...  I can give you the numerical answer here (give me an $x$ I'll tell you the output) but this is not what is wanted.

$$
f(x) = y_1\frac{(x-x_2)\cdots (x-x_n)}{(x_1-x_2)\cdots (x_1-x_n)}+\cdots +y_n\frac{(x-x_1)\cdots (x-x_{n-1})}{(x_n-x_1)\cdots (x_n-x_{n-1})}
$$

So here the answer is

$$
f(x) =5\frac{(x-2)(x-3)(x+4)}{(-2)(-3)(4)}+6\frac{x(x-3)(x+4)}{2(-1)(6)}-6\frac{x(x-2)(x+4)}{3(3-2)(3+4)}+7\frac{x(x-2)(x-3)}{-4(-2)(-1)}
$$

I don't find this very satisfying but I did it!